torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py

```diff
@@ -374,6 +374,361 @@ def collect_bw_donated_buffer_idxs(
+            elif node.args[1].startswith("___backward"):
+                num_fw_inputs = new_hop_graphs[identifier].old_num_fw_inputs
+                assert num_fw_inputs is not None
+                num_fw_outputs = new_hop_graphs[identifier].old_num_fw_outputs
+                assert num_fw_outputs is not None
+
+                # Step 1) - Get the `joint_hop_gm`. As mentioned earlier, the
+                # backward graph is the joint graph.
+                joint_hop_gm = hop_gm
+
+                # Prepare the graph for the partitioner
+                joint_hop_gm = prepare_for_partitioner(
+                    joint_hop_gm, num_fw_inputs, num_fw_outputs
+                )
+
+                # Step 2) and 3) - Run joint graph passes and partitioner
+                new_fw_hop_gm, new_bw_hop_gm = aot_config.partition_fn(
+                    joint_hop_gm, [], num_fwd_outputs=num_fw_outputs
+                )
+
+                # Save the new forward and backward graph modules
+                new_hop_graphs[identifier].new_fw_hop_gm = new_fw_hop_gm
+                new_hop_graphs[identifier].new_bw_hop_gm = new_bw_hop_gm
+
+                # Save the number of symints and saved tensors
+                new_fw_out_nodes = new_fw_hop_gm.graph.find_nodes(op="output")[0].args[
+                    0
+                ]
+                extra_outputs = new_fw_out_nodes[num_fw_outputs:]
+                symint_outputs = [n for n in extra_outputs if is_sym_node(n)]
+
+                new_hop_graphs[identifier].new_num_sym_nodes = len(symint_outputs)
+                new_hop_graphs[identifier].new_num_saved_nodes = len(
+                    extra_outputs
+                ) - len(symint_outputs)
+
+                new_hop_graphs[identifier].partitioning_done = True
+
+    if not new_hop_graphs:
+        return joint_gm
+
+    # Step 3) Restitch the new fw and bw graphs back into the main graph.
+    #
+    # This is a very mechanical process. There are a quite of pieces that we
+    # need to connect together to make it work. Lets try to understand the
+    # problem statement first.
+    #
+    # For the forward graph, the signature of the old_fw_hop_gm is
+    #   inputs - (*primals)
+    #   outputs - (*fw_outs)
+    # Now the signature of the new_fw_hop_gm is
+    #   inputs - (*primals)     -- This is same
+    #   outputs - (*fw_outs, *saved_tensors)    - This is different
+    # At a high level, this is an easy transformation, in the new graph we just
+    # have to replace the old_fw_hop_gm with the new_fw_hop_gm. Everything else
+    # falls into place, because the input signature (i.e. args) is same. And
+    # even though output signature is different, fw_outs are still at the same
+    # indexes as before. So the forward of the `joint_gm` works nicely.
+    #
+    # Now, lets look at the backward hop graph. Old signature
+    #   inputs - (*primals, *tangents)
```

Comment:
I think in particular this is where Elias's rng changes will mess with `invoke_subgraph` - running the min cut partitioner on each inner joint graph will update `num_primals` to also include any new generator arguments, that we need to thread properly

Reply:
True :(